### PR TITLE
[RFC] Fix os_file_exists to not approve directories

### PR DESCRIPTION
**Pull request by danobi**
As described in issue #4939

### CODE REVIEW

**Comment by ZyX-I** [comment]
This will break `vim_rename` and who knows what else. Proposed fix was to fix the name, not behaviour.

**Comment by danobi** [comment]
@ZyX-I Sorry, I misunderstood. 

How does this look now? I renamed `os_file_exists` to `os_path_exists` and updated all the call sites. I also created a function called `os_is_reg_file` that checks if a path points to a regular file, since it looked like that function didn't exist already.

**Comment by ZyX-I** [comment]
@danobi It is not a library for third-party uses, `os_is_reg_file` is not needed until it is used. If you want to write tests for it, I have nothing against, but so far this looks like a waste of time.

I would also suggest adding tests regarding what `os_path_exists` returns for directory, symbolic link and broken symbolik link (latter is not created AFAIK, but is easy to: `lfs.link('xxx_unexistent_file_xxx', 'unit-test-directory/test_broken_link.file', true)`).

**Comment by danobi** [comment]
@ZyX-I Does this look reasonable?

**Comment by ZyX-I** [comment]
This PR LGTM (as its purpose is renaming `os_file_exists` if I am not mistaking), but apparently `vim_rename` has a bug: basically when it uses `os_file_exists` (`os_path_exists` now) it wants to check whether it may use some location to copy file to (i.e. it checks whether location is *occupied*). `os_path_exists` will return `false` for broken symbolic link which may either delete that broken symlink (most likely) or fail without trying any other possible temporary locations (I actually do not know condition under which behaviour will be exactly this). Though Python’s `os.path.exists` does the same thing for broken symlinks (intentional, such behaviour is described in its documentation). Not sure whether this needs to be fixed.

**Comment by ZyX-I** [comment]
Do not forget to silence linter though.

**Comment by justinmk** [comment]
QB is failing because of me, ignore it. I am looking at it now.

File: src/nvim/fileio.c [hunk 1]
```diff
@@ -4514,9 +4515,14 @@ int vim_rename(char_u *from, char_u *to)
...
     STRCPY(tempname, from);
-    for (n = 123; n < 99999; ++n) {
-      sprintf((char *)path_tail(tempname), "%d", n);
-      if (!os_file_exists(tempname)) {
+    for (n = 123; n < 99999; n++) {
+      // the 'n' in snprintf is calculated by subtracting the difference
+      // between the tail pointer and the start of the array from the
+      // total buffer length
+      char * tail = (char *)path_tail(tempname);
+      snprintf(tail, (MAXPATHL + 1) - (tail - (char *)tempname - 1), "%d", n);
```
Comment (justinmk) [inline src/nvim/fileio.c]: the comment is too long, and the mention of "n" is confusing given the presence of the `n` parameter which is unrelated to the 'n' mentioned in the comment. Better to leave the comment out, it mostly restates what the code is doing.

**Comment by danobi** [comment]
Done. Should I be worried about the travis builds? It looks like the failing builds are timing out.

**Comment by justinmk** [comment]
the hanging macOS builds are unrelated, dealing with them elsewhere. Please squash and will merge!

**Comment by danobi** [comment]
Squashed

**Comment by justinmk** [comment]
Thanks! Anything in the [current milestone](https://github.com/neovim/neovim/milestone/10) would help a lot. The [static analysis](https://neovim.io/doc/reports/clang/) issues also need to be fixed.

**Pull request merged**